matminer/featurizers/function.py

```diff
@@ -73,15 +75,53 @@ def exp_dict(self):
                                                    self.combo_function))
              for n in range(1, self.multi_feature_depth+1)])
 
+    def featurize(self, *args):
+        """
+        Main featurizer function, essentially iterates over all
+        of the functions in self.function_list to generate
+        features for each argument.
+
+        Args:
+            *args: list of numbers to generate functional output
+                features
+
+        Returns:
+            list of functional outputs corresponding to input args
+        """
+        return list(self._exp_iter(*args, postprocess=self.postprocess))
+
+    def feature_labels(self):
+        """
+        Returns:
+            Set of feature labels corresponding to expressions
+        """
+        return self._feature_labels
+
+    def fit(self, X, y=None, **fit_kwargs):
+        """
+        Sets the feature labels.  Not intended to be used by a user,
+        only intended to be invoked as part of featurize_dataframe
+
+        Args:
+            X ([str]): list of feature labels
+
+        Returns:
+            Set of feature labels corresponding to expressions
+        """
+        self._feature_labels = self.generate_string_expressions(X)
```

Comment:
X should be a list of entries, and a not a list of column names.

I'd imagine the column names could be passed to this argument a few ways:
1. An optional keyword argument
2. The code detects if X is a dataframe and, if so, uses the column labels.

If the code does not detect label names, we could always make them up (e.g., "x0", "x1").

My reasoning for these suggestions are to make this featurizer follow a similar style to other TransformerMixin classes (which take data as input, not labels), and to allow people to  easily use this class without dataframes (I want to do that)

Reply:
Agree and prefer 2.  I tried 1 in another PR, and my initial approach to that was pretty unwieldy.  Latest commits have this.